### ISSUE TITLE
Detecting crashes of the Pyodide WASM environement

### ISSUE THREAD

**Issue by rth** [iodide-project/iodide#994]
It would be very useful if Iodide could detect when Pyodide WASM environment crashes (e.g. after a segfault), and warn the user. 

For instance, running the simple code snippet from https://github.com/iodide-project/pyodide/issues/200 in the demo notebook https://iodide.io/pyodide-demo/python.html produces a segfault which crashes Pyodide, meaning that any subsequent command will also fail.

This situation is very confusing for users, because Iodide still works in this scenario. So the user gets some error in one notebook cell, and than thinking that maybe something was wrong in the provided code, they would try to change the input code, but irrespective of the input the cell would still fail. More frustrating, even other cells that used to work now will also fail. The only solution is to restart the page.

Even if though segfaults shouldn't happen, it's still possible to run into one, particularly when people build some custom packages. I'm not sure if running out of memory can also produce somewhat similar outcome.

It would help user experience to have some mechanism in Iodide to warn the user when this happens. For instance, one could imagine running some sanity test (e.g executing `1+1`) before running the actual code from the cell.

In Jupyter this is somewhat similar to the "kernel died" situation. I'm not sure if it's possible/desirable to recover from such situation, without restarting the page though.

cc @mdboom

**Comment by mdboom** [comment]
This is definitely in the grey area of responsibility between iodide and pyodide.  A full solution probably involves fixes at both ends, so I'm find with leaving this bug in this tracker.

In the segfaults you've seen, do they always result in `abort` being called?  I wonder if we could override `abort` to set a flag, which we could check on subsequent calls to `runPython` to display the message "Python interpreter crashed" or something...

**Comment by rth** [comment]
Yes, all those I saw, included an `abort`. There is a [`Module.onAbort` callback](https://github.com/kripken/emscripten/blob/906e14e4173e00b47ed931f993c837eb7c6f649a/site/source/docs/api_reference/module.rst#affecting-execution) BTW.

**Comment by mdboom** [comment]
Perfect.  It seems like that might be a good first solution.

**Comment by hamilton** [comment]
@mdboom do you have any thoughts on how we might bubble up / handle these kinds of errors in Iodide? I suppose ideally these types of errors would not require much more on the language plugin maker, and iodide could just (more or less) automatically detect `abort` and give the option to reload the environment (or maybe just the language itself).

**Comment by mdboom** [comment]
From a UX perspective, I think as a first pass, it should pop up a modal saying that the language plugin has crashed, save the page (automatically) and then reload the page.  It will be complicated to try to throw out the old language environment and start a new one without reloading.

Maybe Iodide could pass a callback to the language plugin that the language plugin is responsible for calling when the language plugin crashes?  I think that's the most generic thing, since different languages are likely to crash in different ways (some will be implemented in Wasm, others in Javascript, for example), and Iodide wouldn't be able to detect all cases.

**Comment by hamilton** [comment]
Right. Well, I suppose another approach would be to have a "reload eval environment" action that just reloads the frame (thus "restarting the kernel").

**Comment by mdboom** [comment]
That's true.  Reloading just the iframe is probably enough.

**Comment by bcolloran** [comment]
filed #998 for this
